thrust/examples/run_length_decoding_single_pass.cu

```diff
@@ -0,0 +1,169 @@
+  {}
+
+  template <typename ValueType>
+  __host__ __device__ CountType operator()(run<ValueType, CountType> r) const
+  {
+    cuda::std::size_t end = cuda::minimum()(r.offset + r.count, out_size);
+    thrust::fill(thrust::seq, out + r.offset, out + end, r.value);
+    if (r.run_id == runs_size) // If we're the last run, write the expanded size.
+    {
+      *expanded_size = end;
+    }
+    return end;
+  }
+};
+
+template <typename ValueIterator, typename CountIterator, typename OutputIterator, typename CountType>
+OutputIterator run_length_decode(
+  ValueIterator values, CountIterator counts, CountType runs_size, OutputIterator out, CountType out_size)
+{
+  using ValueType = typename ValueIterator::value_type;
+
+  // Zip the values and counts together and convert the resulting tuples to a named struct.
+  auto runs = thrust::make_transform_iterator(
+    thrust::make_zip_iterator(values, counts), [] __host__ __device__(thrust::tuple<ValueType, CountType> tup) {
+      return run{thrust::get<0>(tup), thrust::get<1>(tup)};
+    });
+
+  // Allocate storage for the expanded size. If we were using CUB directly, we could read this
+  // from the final ScanTileState.
+  thrust::device_vector<CountType> expanded_size(1);
+
+  // Iterator that writes the expanded sequence to the output and discards the actual scan results.
+  auto expand_out = thrust::make_transform_output_iterator(
+    thrust::make_discard_iterator(), expand(out, out_size, runs_size, expanded_size.begin()));
+
+  // Scan to compute output offsets and then write the expanded sequence.
+  thrust::inclusive_scan(thrust::device, runs, runs + runs_size, expand_out);
+
+  // Calculate the end of the output sequence.
+  return out + expanded_size[0];
+}
+
+int main()
+{
+  using ValueType = cuda::std::int32_t;
+  using CountType = cuda::std::size_t;
+
+  CountType size   = 8192;
+  CountType repeat = 4;
+
+  thrust::device_vector<ValueType> values(size);
+  thrust::sequence(thrust::device, values.begin(), values.end(), 0);
+
+  thrust::device_vector<CountType> counts(size);
+  thrust::fill(thrust::device, counts.begin(), counts.end(), repeat);
+
+  // Print first 32 elements of compressed input
+  thrust::host_vector<ValueType> values_h(32);
+  thrust::host_vector<CountType> counts_h(32);
+  thrust::copy(values.begin(), values.begin() + 32, values_h.begin());
+  thrust::copy(counts.begin(), counts.begin() + 32, counts_h.begin());
+
+  std::cout << "Compressed input (first 32 runs):" << std::endl;
+  for (CountType i = 0; i < 32; ++i)
+  {
+    std::cout << "(" << values_h[i] << "," << counts_h[i] << ")";
+    if (i < 31)
+    {
+      std::cout << " ";
+    }
+  }
+  std::cout << std::endl << std::endl;
+
+  thrust::device_vector<ValueType> output(size * repeat);
+
+  auto out_end = run_length_decode(values.begin(), counts.begin(), values.size(), output.begin(), output.size());
+
+  if (static_cast<CountType>(out_end - output.begin()) != size * repeat)
+  {
+    throw int{};
```

Comment:
Important: let's not show bad practices like throwing exceptions not derived from `std::exception`. Since this is in `main`, can we just print an error message and return something nonzero?

Applies one more time below.